src/VisualStudio/Next/Remote/ServiceHubRemoteHostClient.WorkspaceHost.cs

```diff
@@ -0,0 +1,163 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Remote;
+using Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem;
+using Microsoft.VisualStudio.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Remote
+{
+    internal partial class ServiceHubRemoteHostClient
+    {
+        private class WorkspaceHost : ForegroundThreadAffinitizedObject, IVisualStudioWorkspaceHost, IVisualStudioWorkingFolder
+        {
+            private readonly VisualStudioWorkspaceImpl _workspace;
+
+            // We have to capture the solution ID because otherwise we won't know
+            // what is is when we get told about OnSolutionRemoved.  If we try
+            // to access the solution off of _workspace at that point, it will be
+            // gone.
+            private SolutionId _currentSolutionId;
+
+            // We chain all tasks so that all the messages we send to the remote 
+            // process are serialized.  We don't want to interleave the individual
+            // messages we have for things like registering/unregistering the 
+            // primary solution ID.
+            private readonly object _gate = new object();
+            private Task _lastTask = SpecializedTasks.EmptyTask;
+
+            public WorkspaceHost(VisualStudioWorkspaceImpl workspace)
+            {
+                _workspace = workspace;
+                _currentSolutionId = workspace.CurrentSolution.Id;
+                
+                // Ensure that we populate the remote service with the initial state of
+                // the workspace's solution.
+                EnqueueRegisterPrimarySolutionTask();
+            }
+
+            public void OnAfterWorkingFolderChange()
+            {
+                this.AssertIsForeground();
+                EnqueueRegisterPrimarySolutionTask();
```

Comment:
unfortunately you can't do this, since when this method is returned, vs will assume they can delete existing workingfolder. so this API must be synchronous.

there is issue with our delay pushing which I couldn't figure out how to deal with, but in real life, it wasn't actually issue. (no one open solution and rename solution right away before solution load is finished)

but this will make VS not possible to delete old .vs folder when user renamed solution.